domain/secret/state/state.go

```diff
@@ -2459,3 +2466,165 @@ FROM secret_revision sr`, selectStmt)
+
+	selectRevisionParams := []any{secretID{
+		ID: uri.ID,
+	}}
+
+	var revFilter string
+	if len(revs) > 0 {
+		revFilter = "\nAND revision IN ($revisions[:])"
+		selectRevisionParams = append(selectRevisionParams, revisions(revs))
+	}
+
+	selectRevsToDelete := fmt.Sprintf(`
+SELECT uuid AS &revisionUUID.uuid
+FROM   secret_revision
+WHERE  secret_id = $secretID.id%s
+`, revFilter)
+	selectRevisionStmt, err := st.Prepare(selectRevsToDelete, secretID{}, revisionUUID{}, revisions{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	countRevisions := `SELECT count(*) AS &M.count FROM secret_revision WHERE secret_id = $secretID.id`
+	countRevisionsStmt, err := st.Prepare(countRevisions, secretID{}, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteRevisionExpire := `
+DELETE FROM secret_revision_expire WHERE revision_uuid IN ($revisionUUIDs[:])`
+
+	deleteRevisionContent := `
+DELETE FROM secret_content WHERE revision_uuid IN ($revisionUUIDs[:])`
+
+	deleteRevisionValueRef := `
+DELETE FROM secret_value_ref WHERE revision_uuid IN ($revisionUUIDs[:])`
+
+	deleteRevision := `
+DELETE FROM secret_revision WHERE uuid IN ($revisionUUIDs[:])`
+
+	deleteRevisionQueries := []string{
+		deleteRevisionExpire,
+		deleteRevisionContent,
+		deleteRevisionValueRef,
+		deleteRevision,
+	}
+
+	deleteRevisionStmts := make([]*sqlair.Statement, len(deleteRevisionQueries))
+	for i, q := range deleteRevisionQueries {
+		deleteRevisionStmts[i], err = st.Prepare(q, revisionUUIDs{})
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if isLocal, err := st.checkExistsIfLocal(ctx, tx, uri); err != nil {
+			return errors.Trace(err)
+		} else if !isLocal {
+			// Should never happen.
+			return secreterrors.SecretIsNotLocal
+		}
+
+		result := []revisionUUID{}
+		err := tx.Query(ctx, selectRevisionStmt, selectRevisionParams...).GetAll(&result)
+		if err != nil {
+			return errors.Annotatef(err, "selecting revision UUIDs to delete for secret %q", uri)
+		}
+
+		toDelete := make(revisionUUIDs, len(result))
+		for i, r := range result {
+			toDelete[i] = r.UUID
+		}
+		for _, stmt := range deleteRevisionStmts {
+			err = tx.Query(ctx, stmt, toDelete).Run()
+			if err != nil {
+				return errors.Annotatef(err, "deleting revision info for secret %q", uri)
+			}
+		}
+
+		countResult := sqlair.M{}
+		err = tx.Query(ctx, countRevisionsStmt, selectRevisionParams[0]).Get(&countResult)
+		if err != nil {
+			return errors.Annotatef(err, "counting remaining revisions for secret %q", uri)
+		}
+		count, _ := strconv.Atoi(fmt.Sprint(countResult["count"]))
+		if count > 0 {
+			return nil
+		}
+		// No revisions left so delete the secret.
+		return st.deleteSecret(ctx, tx, uri)
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
+
+func (st State) deleteSecret(ctx context.Context, tx *sqlair.TX, uri *coresecrets.URI) error {
+	deleteSecretRotation := `
```

Comment:
We can merge these into same SQL statement, then no need to prepare and Run in a loop.

Reply:
Unless I am mistaken SQLite does not support multiple SQL statements. You have to do it one at a time. And for SQLite this is efficient.

Reply:
Right, just confirmed, sqlite C lib only executes the 1st statement.